Replace nvm, apt, Homebrew, external Qt install and ad-hoc downloads with Nix package manager

pedropombeiro opened the pull request:
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

### Summary:

[comment]: # (Summarise the problem and how the pull request solves it)

This PR introduces the Nix package manager and in the process:
- removes dependencies on nvm, Homebrew and apt-get;
- removes the Dockerfiles in `/docker` and `scripts/3rd-party/nvm/0.33.11/install.sh`;
- removes the requirement to have Qt/macdeployqt pre-installed and uses its own build;
- removes the need to install Android NDK and uses its own build;
- removes yarn 1.13.0 local install from the repo;
- moves from downloading linuxdeployqt (Linux) and NSIS (Windows) to using its own build in Nix.

Sorry for the size of the PR, but it is a big undertaking at the same time. I've tried to fixup commits as much as possible to keep them as focused as possible and get rid of noise/failed attempts.

Documentation updates are done in https://github.com/status-im/status.im/pull/244.

The Nix scripts are structured as follows:

```
- /default.nix
  - /scripts/lib/setup/nix
    - desktop/default.nix
      - appimagetool/default.nix
      - linuxdeployqt/default.nix
      - windows/default.nix
        - conan/default.nix
        - nsis/default.nix
    - global-node-packages/generate-nix.sh
      - output/default.nix
    - mobile/default.nix
      - android-ndk/default.nix
```

### Done

- [x] Replace existing tools with Nix equivalents (Conan, Node, NVM, yarn, Leiningen, OpenJDK, Maven, Python, etc.);
- [x] Install react-native using Nix instead of npm;
- [x] Integrate Qt5 Nix package (Linux and macOS);
- [x] Fix up desktop app libraries on both Linux and macOS;
- [x] Remove yarn 1.13.0 release from repo;
- [x] Replace NSIS apt install with custom Nix recipe;
- [x] Replace Android NDK, `appimagetool` and `linuxdeployqt` downloads with custom Nix recipes;
- [x] Move deployment helper tools used for generating Linux Desktop app to Nix recipes;
- [x] Split `/default.nix` to task-based Nix expressions. E.g. we only need to install Conan and NSIS when running `make release-windows-desktop`;

### TODO list (as of 28.02.2019)

- Blockers:
  - [x] Fix rpaths of 3rd party libraries to not point to /nix/store;
  - [x] Fix Linux CI issue with `qtkeychain` (fixing locally breaks CI build and vice-versa);
  - [x] Fix macOS binary cache;
  - [x] Merge `react-native-keychain` [PR](https://github.com/status-im/react-native-keychain/pull/10);
  - [x] Fix NSIS Nix package so it relocates libraries that are installed in system directories, such as `/usr/share/nsis/Stubs/zlib`;
  - [x] Migrate CI scripts to rely on Nix;
  - [x] Fix debug builds (e.g. Desktop);
- Not a blocker (can be done on subsequent PRs):
  - [x] Fix issue where first `make shell` (which invokes `setup` script, which invokes `make prepare-mobile`) fails on `make prepare-mobile` due to `sha3` package not finding `node.h`;
  - [x] Make upstream PR on nixpkgs to include webview on `qt5.full` package;
  - [x] Fix missing icons in macOS builds (missing `imageformats` and `iconengines` plugins in bundle);
  - [ ] Optimization: Share /nix/store with host so that each job doesn't need to fetch the whole Nix store contents;
  - [ ] Optimization: Avoid downloading Android SDK every time on CI builds;
  - [x] Use NSIS from nixpkgs once [upstream PR](https://github.com/NixOS/nixpkgs/pull/56990) is merged;
  - [ ] `nix-cache.status.im` is not being used in NixOS;
  - [ ] Patch Android NDK binaries to use Nix's ld interpreter and Nix dependencies, so that it can run on NixOS (ideally just build from source);
  - [ ] Standardize on Qt 5.11.3.

### Review notes (optional):
<!-- (Specify if something in particular should be looked at, or ignored, during review) -->

Ideally, test this on your own machine - better to do it know and find any issues than to find out when this is on `develop`.
Running this branch should have no impact on the system and be complete side-by-side capable. It will download the Android SDK and NDK to a location we completely control `~/.status/Android/Sdk`.

There is a subtle change in how we do things with respect to .TOOLVERSIONS. So far we've been installing tools based on the versions listed in .TOOLVERSIONS. Since now the tool versions are decided by what is present in the Nixpkgs snapshot we're using, most of the tools in .TOOLVERSIONS can be delisted. They've only been kept so that we can build Docker images based on them.

### Testing notes (optional):

We should give special attention to testing Status Desktop, especially on macOS Mavericks, since Nix normally targets SDK 10.10 instead of 10.9 as has been the case so far for us.

### Next steps:

- Optimize CI builds  so they share the Nix store;
- Distribute status-go builds through Nix;
- Encapsulate `StatusIm-Windows-base-image.zip` and `StatusImAppImage_20181208.zip` as Nix packages;
- Leverage [yarn2nix](https://github.com/moretea/yarn2nix) and other similar tools (e.g. for Ruby gems);
- With Conan being installed by Nix, and Qt for Windows being built and installed by Conan, having to manually download Qt for Linux/macOS sticks out like a sore thumb. I think we should make take our current Windows Qt script and make a Conan package for Linux/macOS (even better would be to do everything in Nix and discard Conan too :-) but the time investment is not the same);
- Install Android SDK, build-tools and platform-tools using Nix, so we don't pollute the user's system and we have a pure Nix environment.

status: ready <!-- Can be ready or wip -->

Commit:
Migrate from installing tooling the traditional way to Nix

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 477bec55 | [#1](https://ci.status.im/job/status-react/job/prs/job/ios-e2e/job/PR-7506/1/) | 2019-02-14 15:37:40 | ~22 sec | `ios-e2e` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/ios-e2e/job/PR-7506/1/consoleText) |

cammellos commented on .mailmap:
```diff
@@ -0,0 +1,16 @@
+Andrea Maria Piana <cammellos@cammelloss-MacBook-Pro.local>
```
eemh, not sure this is meant to be here?

rasom replied:
why not? :)

pedropombeiro replied:
@cammellos This was just to make our names uniform across the repo. In this case there were commits that you've made to the repo using that email address. Do you prefer not to have that?

jakubgs replied:
Then maybe we should use status emails?

cammellos replied:
didn'd even know what a mailmap file was :D , goddamn spelling errors on hostname!

Commit:
Migrate Desktop build environment to leverage Nix

- Add Conan to default.nix
- Integrate Qt5 Nix package (with forked repo for new QtWebView module)
- Fix rpaths in macOS and Linux
- Add Nix expressions to install nsis, appimagekit and linuxdeployqt

pedropombeiro commented:
Added Conan and jq to Nix script and added Next steps section to OP.

jakubgs commented on Makefile:
```diff
@@ -40,17 +40,29 @@ __toolversion = $(shell $(GIT_ROOT)/scripts/toolversion $(1))
+endif
 
 prepare-desktop: ##@prepare Install desktop platform dependencies and prepare workspace
+ifndef IN_NIX_SHELL
```
You can make a target called `nix-shell-check` that would have this `ifndef` and re-use it in other targets.
```Makefile
nix-shell-check:
ifndef IN_NIX_SHELL
	@echo "Please run '$(MAKE) shell' first"
else
```
And then just add it as first dependent target.

pedropombeiro replied:
But that won't prevent the invoked target (`prepare-desktop` in this case) from running to completion, right?

jakubgs replied:
It will if you add `exit 1`. Or use `error` like I do here:
https://github.com/status-im/status-go/blob/72906ac6553869f6ac6df4e294a91c184cb320a4/_assets/compose/mailserver/Makefile#L23-L25

pedropombeiro commented:
Looks like this is building on some mac agents but not on others.

pedropombeiro commented:
Found a blocker for this PR. Although I can build Status Desktop for Windows, I can't build it for Linux:

```
-- Qt root directory: /home/pedro/opt/Qt/5.11.2/gcc_64
CMake Error at /home/pedro/opt/Qt/5.11.2/gcc_64/lib/cmake/Qt5Gui/Qt5GuiConfigExtras.cmake:9 (message):
  Failed to find "GL/gl.h" in "/usr/include/libdrm".
Call Stack (most recent call first):
```

Looks like that path is hardcoded in `Qt5GuiConfigExtras.cmake`, but this file is somehow not included when running outside of a Nix shell. I compared `CMakeCache.txt` files from both scenarios and on the Nix one some `Qt5_*DIR`s variables are missing (`Qt5Concurrent_DIR`, `Qt5DBus_DIR`, `Qt5Widgets_DIR` and `Qt5_DIR`), as well as `GO_EXECUTABLE`.

As a parallel approach, I tried using the nix `qt5.full` package and that looks much more promising but I noticed 4 other problems:
- `QtConcurrent` is not included in `qt5.full`. Will look into replacing calls with `std::async` which is anyway a good improvement in and of itself;
- I get a cgo build error:
```
make[4]: Entering directory '/home/pedro/src/github.com/status-im/status-react/desktop/modules/react-native-status/desktop/StatusGo/src/github.com/status-im/status-go'
make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Building static library...
# github.com/status-im/status-go/vendor/github.com/karalabe/hid
cgo-dwarf-inference:2:8: error: enumerator value for '__cgo_enum__0' is not an integer constant
cgo-dwarf-inference:10:2: error: initializer element is not constant
cgo-dwarf-inference:10:2: note: (near initialization for '__cgodebug_ints[0]')
```

  UPDATE: looks like a known issue: https://github.com/NixOS/nixpkgs/issues/25959. Tried installing the `gcc6` package and status-go was able to build.

- `qt5.full` is not installable on macOS, due to a recipe bug: https://github.com/NixOS/nixpkgs/issues/55892.
- `linuxdeployqt` doesn't know about Nix packages library paths, so when going about removing rpaths, it complains about missing libraries.

So still some ways to go, I guess.

ghost commented:
`desktop_files/yarn.lock` changed. Pinging @jakubgs, @mandrigin, and @corpetty

ghost commented:
`desktop_files/yarn.lock` changed. Pinging @jakubgs, @mandrigin, and @corpetty

jakubgs commented on scripts/build-desktop.sh:
```diff
@@ -312,28 +339,32 @@ function bundleLinux() {
+  # Ensure the binary isn't using the interpreter in Nix's store
+  patchelf --set-interpreter /lib64/ld-linux-x86-64.so.2 "$usrBinPath/Status"
+  # Tell linuxdeployqt about all the different lib folders in Nix's store 
+  local all_deps=($(get_recursive_dependencies "$usrBinPath/Status"))
```
Clever use of `ldd` to find necessary `LD_LIBRARY_PATH` contents to compile them in with `deployqt`. Very cool!

ghost commented:
`desktop_files/yarn.lock` changed. Pinging @jakubgs, @mandrigin, and @corpetty

jakubgs commented on modules/react-native-status/desktop/rctstatus.cpp:
```diff
@@ -20,7 +20,7 @@
 #include <QDir>
 #include <QStandardPaths>
-#include <QtConcurrent>
+#include <future>
```
Can you explain this change? Was nix unable to install `QtConcurrent`, or is this just some side change?

pedropombeiro replied:
Yeah, Nix's `qt5.full` package doesn't seem to include `QtConcurrent`, but even with that aside, I think it's not ideal to add a dependency when the standard library already provides the same functionality. I.e. even if this PR doesn't go forward, I'd still want to include this change in another targeted PR.

veprbl replied:
QtConcurrent is part of `qt5.qtbase` which is included in `qt5.full`

jakubgs replied:
That was my understanding as well.

pedropombeiro replied:
What made me look into an alternative was this error right after I brought in the `qt5.full` Nix package:
```
CMake Error at /nix/store/d28n5vfavdlg0gd43cgd52hcyhm96xaz-qt-full-5.12.0/lib/cmake/Qt5Core/Qt5CoreMacros.cmake:372 (message):
  Can not use "Concurrent" module which has not yet been found.
Call Stack (most recent call first):
  /home/pedro/src/github.com/status-im/status-react/node_modules/react-native/ReactQt/runtime/src/CMakeLists.txt:259 (qt5_use_modules)
```
I could take some time and delve into it, but the truth is that the reason we import QConcurrent is not strong enough when there is an equivalent solution with `std::async`.

pedropombeiro replied:
Huh, turned out to be a missing import in one file (no idea why this wasn't a problem pre-Nix). I'll revert the removal of QtConcurrent for now, to make the PR as simple as possible.

ghost commented:
`desktop_files/yarn.lock` changed. Pinging @jakubgs, @mandrigin, and @corpetty

ghost commented:
`desktop_files/yarn.lock` changed. Pinging @jakubgs, @mandrigin, and @corpetty

jakubgs commented on scripts/lib/setup/installers.sh:
```diff
@@ -1,406 +1,64 @@
+      if is_linux && [ -f ~/.bashrc ]; then
+        # For some reason, new terminals are not started as login shells, so .profile and .bash_profile are not sourced.
+        # To get around it, we add Nix initialization to .bashrc as well, if it exists
+        echo "if [ -e ${HOME}/.nix-profile/etc/profile.d/nix.sh ]; then . ${HOME}/.nix-profile/etc/profile.d/nix.sh; fi # added by `make setup` Status script" >> ~/.bashrc
```
This is a bug. By doing backticks inside of double quotes in which you have `make setup` you cause an infinite loop of `make setup` being called by `make setup` being called by `make setup` being called by `make setup` being called...

pedropombeiro replied:
Good catch! We don't need those backticks there

jakubgs commented:
I can't seem to finish `nix-build` successfully:
```
[nix-shell:~/work/status-react]$ nix-build
these derivations will be built:
  /nix/store/v4yhm1gmd2gcv31n93l51hyr3gnjay6y-env.drv
building '/nix/store/v4yhm1gmd2gcv31n93l51hyr3gnjay6y-env.drv'...
unpacking sources
variable $src or $srcs should point to the source
builder for '/nix/store/v4yhm1gmd2gcv31n93l51hyr3gnjay6y-env.drv' failed with exit code 1
error: build of '/nix/store/v4yhm1gmd2gcv31n93l51hyr3gnjay6y-env.drv' failed
```
Not sure what to make of it.

pedropombeiro commented:
@jakubgs try `nix-build -A env` (passing the derivation name in `default.nix`)

jakubgs committed:
add jenkinsfile

Signed-off-by: Jakub Sokołowski <jakub@status.im>

Commit:
Change shebang in bash scripts

Commit:
Fix problem building `sha3` node module in `make setup`

Commit:
Add Android NDK Nix expression

Commit:
Implement target-dependent installing of Nix dependencies

Commit:
Migrate remaining CI builds to Nix

Commit:
Remove build Dockerfiles and NVM script

Commit:
unfinished: Bundle all Linux dependencies so that app can run on all distros

PombeirP commented:
@annadanchenko it would be great if we could start doing some focused preliminary testing this PR ahead of reviewing, as this might uncover more work that needs to be done before a final review. Namely some platforms that could be problematic such as macOS, where we need to check that the work I did to maintain compatibility with the 10.9 SDK was successful (i.e. are we able to run the app on the same OS versions as before). Linux is also a candidate for this type of testing, however, the CI builds are not yet ready (i.e. although the app runs fine on 16.04, the generated AppImage on CI isn't yet ready to run on 16.04). I will let you know when Linux can be tested.

PombeirP commented:
https://github.com/probonopd/linuxdeployqt/issues/356

jarradh commented:
So I am on NixOS and I followed the new docs.

https://github.com/status-im/status.im/pull/244/files#diff-b0094888c5050c10449718b8e462c289R19

1) it required me to run `make shell` before `make setup`


2) it asks for sudo password before downloading Android SDK
```
Checking prerequisites...

[sudo] password for user: 
sysctl: cannot stat /proc/sys/kernel/unprivileged_userns_clone: No such file or directory
Installing requirements...

Downloading Android SDK.
```

3) I got as far as failed on `make prepare-desktop` on 
```
[nix-shell:~/experiments/status-react]$ make prepare-desktop
make[1]: Entering directory '/home/j/experiments/status-react'
scripts/prepare-for-platform.sh desktop
Finished!
Creating link: package.json -> desktop_files/package.json.orig
Creating link: yarn.lock -> desktop_files/yarn.lock
Creating link: VERSION -> desktop_files/VERSION
Creating link: .babelrc -> desktop_files/.babelrc
Finished!
yarn install --frozen-lockfile
yarn install v1.13.0
[1/4] Resolving packages...
[2/4] Fetching packages...
[######################################################################################################----------------------------] 976/1245^C^Z
[1]+  Stopped                 make prepare-desktop

[nix-shell:~/experiments/status-react]$ make shell
make[1]: Entering directory '/home/j/experiments/status-react'
Nix shell is already active
make[1]: Leaving directory '/home/j/experiments/status-react'

[nix-shell:~/experiments/status-react]$ make prepare-desktop
make[1]: Entering directory '/home/j/experiments/status-react'
scripts/prepare-for-platform.sh desktop
Finished!
Creating link: package.json -> desktop_files/package.json.orig
Creating link: yarn.lock -> desktop_files/yarn.lock
Creating link: VERSION -> desktop_files/VERSION
Creating link: .babelrc -> desktop_files/.babelrc
Finished!
yarn install --frozen-lockfile
yarn install v1.13.0
[1/4] Resolving packages...
[2/4] Fetching packages...
error Command failed.
Exit code: 1
Command: git
Arguments: pull
Directory: /home/j/.cache/yarn/v4/.tmp/1c68a3e6bea0ee55b6643bc5d40f5939
Output:
There is no tracking information for the current branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> master
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

make[1]: *** [Makefile:66: prepare-desktop] Error 1
make[1]: Leaving directory '/home/j/experiments/status-react'

[nix-shell:~/experiments/status-react]$ 
```

and failed, I stopped it the first time because I thought I missed a step.

Serhy commented:
MasOS build installed (upgraded from few days-old nightly build) on MacOS Mojave v10.14.2:

<img width="1034" alt="Screenshot 2019-03-12 at 12 08 17" src="https://user-images.githubusercontent.com/8749671/54193786-72645400-44c3-11e9-8000-ae52e766aea9.png">

Icons which are not visible:
1) No Home and Profile icons
2) No `+` in the ‘Create/Join new chat’
3) No arrow icon in ‘Send message’ button
4) More chat options icon is not visible (but exist as an element - i.e. I can click it to call for chat options)
5) No icon on ‘Close quote’ icon
6) Public chat names has a space in chat lists
7) No ‘+’ icon the `Start new chat`, `Start Group chat`, `Join Public chat`
8) If user has the custom picture set for the account - it’s not showing in chat/chats list/top chat bar


9) No QR code for user account is shown
<img width="1017" alt="Screenshot 2019-03-12 at 12 09 10" src="https://user-images.githubusercontent.com/8749671/54194123-2fef4700-44c4-11e9-914a-ef991814f4c4.png">


10) Particular views (with the list of chat items) require a double clicks (the first click apparently moves the view onfocus - so next clicks are actionable and select chat element). It happens if the view was UNfocus before. Please see scenarios below for when and where it happens:   

**Scenario #10(a)** 
1) Switch to Profile view
2) Switch to Home view.
3) Click any chat element from left chat list panel
Actual result: Nothing happens
4) But, click one more time on ANY chat item from the list.
Actual result = Expected result: Chat view opened.

**Scenario #10(b)** 
1) Switch to Home view
2) Click ‘+’(add new chat) button -> `Join public chat`
3) Click any element from the default public chats items on the list.
Actual result: Nothing happens
4) But, click one more time on ANY chat item from the list.
Actual result = Expected result: Chat view opened.

**Scenario #10(c)** 
1) Switch to Home view
2) Select any chat
3) Type text message
4) Click ‘Send message’ button
5) Click any chat element from left chat list panel
Actual result: Nothing happens
6) But, click one more time on ANY chat item from the list.
Actual result = Expected result: Chat view opened.
NOTE: it’s all fine in this scenario if to skip 4th step above

churik commented:
@PombeirP 
checked last [build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190311-164155-7180b5-pr.AppImage)  with commit for fixup Linux - can't start as well.

Don't see any changes on windows build compared to recent nightly, so lgtm.

Serhy commented:
more for MacOS build:
<img width="1002" alt="Screenshot 2019-03-12 at 11 43 22" src="https://user-images.githubusercontent.com/8749671/54198478-42bb4900-44cf-11e9-9a90-b0da1ff022d7.png">

11) Issue with displaying of Status logo
12) No `<` (back) button is visible

13) No `<` / `>` icons displayed on About, Help center screens 
<img width="748" alt="Screenshot 2019-03-12 at 14 06 56" src="https://user-images.githubusercontent.com/8749671/54198882-3b486f80-44d0-11e9-81d1-8db77f837163.png">

rasom commented:
Tried branch locally (development):
- successfully built android version of the app 
- can't run iOS version https://gist.github.com/rasom/85c7768cd5d8b07bc3edf0ca0ac709d2

Commit:
Revert last commit and patch linuxdeployqt so that it doesn't require massaging of files

- After decision to not support Ubuntu 16.04, we can drop the code that copied all the dependency tree

PombeirP commented:
> @PombeirP
> checked last [build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190311-164155-7180b5-pr.AppImage) with commit for fixup Linux - can't start as well.

Linux wasn't yet ready to be tested. I just pushed a new version of the branch which can probably be tested.

PombeirP commented:
> Tried branch locally (development):
> 
>     * successfully built android version of the app
> 
>     * can't run iOS version https://gist.github.com/rasom/85c7768cd5d8b07bc3edf0ca0ac709d2

@rasom that's exactly the same error that I saw from an external contributor that reached me on Status, which at least is a good sign meaning that you have the same environment. Not sure why is different on CI and on my laptop that makes it pass. Do you have any idea? I'll try latest branch again on macOS.

PombeirP commented:
@jarradh I hadn't yet tested this on NixOS, so it's bound to have minor problems. I've pushed a new feature/nix branch and also documentation updates which should get rid of the initial problem you saw. Regarding the problem you saw during `make prepare-desktop`, I'll need to resize the VM disk from their Virtualbox pre-made image in order to be able to try that.

PombeirP commented:
@Serhy that is listed in the TODO section as a non-blocker: "Fix missing icons in macOS builds (missing imageformats and iconengines plugins in bundle)"

PombeirP commented:
@jarradh OK, I resized the VM disk and starting from scratch, I was able to run `make desktop` successfully. Only thing I did was:
- `nix-env --install git gnumake`
- clone repo
- `make shell`
- `make prepare-desktop`

PombeirP commented:
@churik current Linux build should work now.

PombeirP commented:
@rasom one thing I noticed in your logs is that you're using the new build system:

https://gist.github.com/rasom/85c7768cd5d8b07bc3edf0ca0ac709d2#file-ios-log-L14

PombeirP commented:
@rasom, one way I was able to unblock that was to patch `local-cli/runIOS/runIOS.js` so that it passes `-UseModernBuildSystem=NO` to `xcodebuild`. Is that something we can patch directly in our fork of `react-native`?

PombeirP commented:
Looks like something got broken in the past 2 weeks on iOS. I just fired up a VM where I'd been able to run `make run-ios` successfully and will compare what changed.

rasom commented:
@PombeirP 
when I run app inside Xcode `clang` commands are executes with `-index-store-path /Users/romanvolosovskyi/Library/Developer/Xcode/DerivedData/StatusIm-hiverhierqkxmigmaofvdskluchw/Index/DataStore`, with `make run-ios` it is `-index-store-path /Users/romanvolosovskyi/clj/status-react/ios/build/Index/DataStore`

rasom commented:
@PombeirP 
> click on File->Project Settings, and there change the value in the Build System dropdown to Legacy Build System

this fixed the issue on my side

PombeirP commented:
@rasom What if you start xcode from the command line while you're inside Nix shell?

PombeirP commented:
In order to fix iOS build, I added a patch for `react-native-cli` that will force using the legacy build system (we can get rid of that when we upgrade to RN 0.57+) and modified Nix to not install the compiler on macOS (use `stdenvNoCC` instead of `stdenv`). We'll just assume that it's there and use whatever version is on the user's machine. We can look into that later on.

jarradh commented:
I use [home-manager](https://github.com/rycee/home-manager) to manage user space both `git` and `gnumake`  are defined in in `home.nix`, it's better than nix-env as it is stateful. Trying to run `nix-env --install git gnumake` anyway causes conflict with git. `git` is also in my `configuration.nix`.

I removed gnumake and git from home.nix and ran `make prepare desktop` and it still fails on the same error in yarn.

I try fresh term session, `git pull` then `git reset --hard` followed by running `make setup` and it asks for sudo perms on Checking prerequisites... (seems this is before downloading android again), run `make prepare-desktop` again and it fails in the same place.

PombeirP commented:
Added item to TODO list, since the fix for debugging mobile iOS (i.e. not installing CC on Darwin) broke the desktop build.

PombeirP commented:
OK, Desktop debug builds should be fixed both MacOS and Linux. I believe we're now ready for a code review. Ping @mandrigin @cammellos @jakubgs @rasom.

PombeirP commented:
> I then tried a fresh term session...and all went well :) it seems the Ctrl+C borked the session, so not robust to interrupts.

Hmm, sounds like the issue is not necessarily with Nix but with npm/yarn, probably would have happened in `develop` branch as well. Should be better once we replace the `yarn install` step with a Nix wrapper.

> I noticed an error during build, unsure if relevant, didn't seem to stop anything, seems to be chugging ahead..

Yes, the same thing happens in `develop`, seems harmless (although we probably shouldn't have merged something in that state to `develop` in the first place).

I've built a VM that mimics the same setup you have and am currently testing `make release-desktop`. Something is clearly wrong there.

Serhy commented:
MacOS build (https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190313-133158-1e748b-pr.dmg) has been smoke tested on macOS Mojave 10.14.3 and confirmed there are no issues apart lost icons reported in https://github.com/status-im/status-react/pull/7506#issuecomment-471949112
* Can upgrade from nightly
* Can create and recover an account
* Can join/create 1-1/public/group chats
* Can message in any type of chat
* Can receive offline inbox history
* Can pair account between MacOS <-> iOS(nightly) builds 
* Can clear history and delete chats
* Can switch mailservers

@PombeirP, we're now ready for a code review from testing perspective

PombeirP commented:
Thanks @Serhy!

PombeirP commented:
@jarradh turns out the bundling error was due to missing `file` command. I'll add that to the desktop dependencies.

PombeirP commented:
@jarradh got to the point where the AppImage is generated, although it doesn't seem to be recognized on NixOS, but that's probably an issue with AppImages and NixOS

PombeirP commented:
@vkjr please try this branch out to make sure your normal Desktop workflows haven't been broken.

cammellos reviewed (approved):
Tested on arch linux 64 clean installation, both mobile (android) and desktop, no issues at all and very smooth, amazing work!

vkjr commented:
@PombeirP, I tried this branch and have issue.
In first terminal window I ran:
`make prepare-desktop`
`make watch-desktop`
In second one:
`npm start`

`npm start` gives error:
```
error: bundling failed: Error: Unable to resolve module `react-native-webview` from `/Users/vkjr/work/projects/github/test-pedro-branch/status-react/index.desktop.js`: Module `react-native-webview` does not exist in the Haste module map
```

I believe this is because `clj-rn.conf.edn` still references `react-native-webview`, whereas `desktop_files/package.json.orig` doesn't.
If I remove reference from `clj-rn.conf.edn`, I can run app successfully.

PombeirP commented:
Thanks for trying the branch @vkjr. I don't see how this is related to Nix - are you saying that the same doesn't happen on `develop`?

vkjr commented:
Nope, I'm not saying this, didn't check the `develop`. Just mentioning the problem that exists in this branch :)

PombeirP commented:
OK, but the goal is to see if this branch is breaking something in changing to Nix. If a problem exists in develop it should be fixed in another PR.

oskarth commented:
@vkjr could you check if the same error occurs on develop? If it does it seems to be a desktop specific issue that we should fix, separately from this piece of work.

vkjr commented:
@oskarth, sure, already making a pr

vkjr commented:
#7736

PombeirP commented:
Created a testing matrix for the Nix branch, please fill in if you've successfully tested a missing/unmarked configuration/distro: https://docs.google.com/spreadsheets/d/1T1GBrVEP_uUActnPy2YEA_9q_tDaVOKAsSr6uaJXD_Q/edit?usp=sharing

churik commented:
@PombeirP Can you please elaborate on what areas can be affected (i.e. 1-1 chats, wallet etc) except of Status Desktop?
I'm just confused with a lot of comments, so want to know what exactly is needed for test team in terms of this PR to move it `To merge` column.

PombeirP commented:
> @PombeirP Can you please elaborate on what areas can be affected (i.e. 1-1 chats, wallet etc) except of Status Desktop?

@churik In terms of functional areas, I don't expect subtle errors to appear. Since this deals almost exclusively with build/deployment, some of the effects I've seen are:
- problems running the app on a different machine than the one it was compiled in (in the case where it is pointing to hardcoded Nix paths in the build machine);
- lack of icons/pictures (as is the case in macOS since the deployment tool is not packaging the SVG Qt plugin);
- failure to log in due to problems in the qtkeychain module;
- problems that happen during building/debugging software (e.g. `make run-*`, `make release-*` commands) in a particular combination (e.g., debugging Android on macOS).

I've caught most (hopefully all) of these issues in the past few weeks, so atm the most important is that:
- e2e tests are passing;
- people can run the builds from CI server on all expected platforms (e.g. the oldest supported version of each platform);
- developers confirm that this hasn't broken anything for them (which is why I sent out the [survey](https://docs.google.com/forms/d/1btFBOw9OUuAW2WyNLOw8XYcZGwS2p_rM6QDvRgEv9CE/edit) and the [test matrix](https://docs.google.com/spreadsheets/d/1T1GBrVEP_uUActnPy2YEA_9q_tDaVOKAsSr6uaJXD_Q/edit?usp=sharing)).

PombeirP commented:
@vkjr did `make run-desktop` work correctly for you?

churik commented:
Tested on 5 platfroms - smoke test that can install, create account and login + exploratory testing (navigating on different views, switching between tabs, joining several public chats).
1) Icons on Mac OS will be fixed separately as far as I understand?
2) Still can't launch Linux build (Ubuntu 18.0.4), [build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190314-233824-968df0-pr.AppImage).
3) Noticed, that there are no platforms shown in https://github.com/status-im/status-react/pull/7506#issuecomment-463673237. Should be fixed here or separately?

On Android, IOS and Windows don't see any differences between nightly <-> PR build.

PombeirP commented:
>     1. Icons on Mac OS will be fixed separately as far as I understand?

Yes. I'll create issues from this PR once it is merged.
 
>     2. Still can't launch Linux build (Ubuntu 18.0.4), [build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190314-233824-968df0-pr.AppImage).

Wasn't aware of this, I'll take a look.

>     3. Noticed, that there are no platforms shown in [#7506 (comment)](https://github.com/status-im/status-react/pull/7506#issuecomment-463673237). Should be fixed here or separately?

Hmm, should be fixed here. Maybe @jakubgs can take a look on Monday if I don't get to it sooner.

PombeirP commented:
@antdanchenko there's a test failing in this PR but there should be no changes in functionality. Is it a flaky test?

PombeirP commented:
@churik OK, I know what's wrong with the Linux build. The AppImage doesn't have the executable bit set (`chmod +x`). I'll fix that in the script.

PombeirP commented:
hmm, it's already doing that (as I expected): https://github.com/status-im/status-react/blob/feature%2Fnix/scripts/build-desktop.sh#L303

Wonder why it is getting lost in the artifacts.

antdanchenko commented:
@PombeirP The test failed because an android service `Messages` has stopped, https://assets.saucelabs.com/jobs/11aba5e64e364c30ba302766f9d72eb7/0001screenshot.png?auth=d49ae277701810477dfa09e403987ec9

churik commented:
@PombeirP just ping me when I need to retest Linux build and 3 from https://github.com/status-im/status-react/pull/7506#issuecomment-473194265 will be fixed as well.
Thanks!

PombeirP commented:
@churik I don't need to do anything on the Linux build. The only thing that it's missing is that you need to enable the executable bit `chmod +x StatusIm-x86_64.AppImage` after you download it. I thought it might be happening only on this PR, but I downloaded a Linux build from another PR and it's the same thing.

churik commented:
@PombeirP I actually did it every time, let me check once more

PombeirP commented:
> @PombeirP I actually did it every time, let me check once more.
> UPD: I did it, I marked file as executable, but still Status doesn't launch
> OS: Ubuntu 18.0.4, [build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190314-233824-968df0-pr.AppImage)
> Video: https://take.ms/zILbDd

@churik that's interesting, it's exactly the same thing I did and it launched fine for me (maybe because I have the Nix build directories locally?)

Can you please try running that AppImage from the command line to see what is output?

PombeirP commented:
@churik can you please paste result of `lsb_release -a` on the machine you're testing on?

churik commented:
@PombeirP 
```status@parallels-vm:~$ Desktop/PR.AppImage 
Desktop/PR.AppImage: /lib/x86_64-linux-gnu/libz.so.1: version `ZLIB_1.2.9' not found (required by /tmp/.mount_PR.AppCe0AKE/usr/bin/../lib/libpng16.so.16)
Desktop/PR.AppImage: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.25' not found (required by /tmp/.mount_PR.AppCe0AKE/usr/bin/../lib/libsystemd.so.0)
Desktop/PR.AppImage: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.26' not found (required by /tmp/.mount_PR.AppCe0AKE/usr/bin/../lib/libsystemd.so.0)
Desktop/PR.AppImage: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.27' not found (required by /tmp/.mount_PR.AppCe0AKE/usr/bin/../lib/libsystemd.so.0)
Desktop/PR.AppImage: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.25' not found (required by /tmp/.mount_PR.AppCe0AKE/usr/bin/../lib/libmount.so.1)
Desktop/PR.AppImage: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.25' not found (required by /tmp/.mount_PR.AppCe0AKE/usr/bin/../lib/libblkid.so.1)
```

PombeirP commented:
@churik hmm, looks like we're building with a 18.04 version that is too new compared to the one you're running. cc @jakubgs.

I'll download a VM of 18.04 that is not a recent one so that I can reproduce and test. I wasn't expecting differences in glibc between patch versions of Ubuntu.

PombeirP commented:
Ping @vkjr regarding https://github.com/status-im/status-react/pull/7506#issuecomment-473022424 (I don't think you answered).

PombeirP commented:
@churik OK, that explains it, we don't support Ubuntu 16 anymore (starting in this branch), so you'll need to test with Ubuntu 18.04.

Serhy commented:
Not sure if below is relevant but what I've tried to run Status in MavericksOSX:

1) In VirtualBox installed OS X Maverics 10.9.1 (and upgraded then to v10.9.5)
2) Download MacOS build from this PR (and Stable MacOS build from https://status.im/stable too)
3) Installed Status (moved app to Applications folder)
4) Launch Status from launchpad
**Actual result:** Next error appears (for both Status versions from 2nd point above)
<img width="395" alt="Screenshot 2019-03-15 at 13 38 27" src="https://user-images.githubusercontent.com/8749671/54444194-e6fcf400-474a-11e9-91ef-4692279ee636.png">

5) Run Status from command line: `./Applications/Status.app/Contents/MacOS/Status`
**Actual result:** Error appear with stacktrace here:  
[MavericksOSStatusError.log](https://github.com/status-im/status-react/files/2971898/MavericksOSStatusError.log)

I also downloaded Telegram, Viber, WhatsUp dmg and installed them too:
- Telegram, as I found out later, does not support OS X less than v10.11
- Slack.app started fine
- WhatsUp crashed on launch with a similar error as Status in above scenario.

I would rather blame Virtual env in my case, but if above errors are anything to you, @PombeirP please let me know.

With MacOS Mojave 10.14 issue with icon lead me to question: how much time may it take to bring back icons in MacOS? I'd say that MacOS with icon issues (https://github.com/status-im/status-react/pull/7506#issuecomment-471949112) is a blocker issue (since it makes MacOS build almost unusable).

mandrigin commented:
> With MacOS Mojave 10.14 issue with icon lead me to question: how much time may it take to bring back icons in MacOS?

@Serhy I would much rather bite the bullet and merge it with this issue than keeping it and rebasing and re-testing because some new regressions might have been introduced by changed *develop* branch...

PombeirP commented:
@Serhy @mandrigin It's very possible that this version of Status doesn't support Mavericks (it looks like Nix's goal isn't to support older versions of OSes), so we need to ensure we're OK with not supporting it (at least until we find a solution) before merging this PR.

vkjr commented:
@pombeirp, yes it worked for me, but I use ‘react-native run-desktop’, not make command

PombeirP commented:
@vkjr I believe the make command used to work, at least a few months ago. It would be nice to have the make command working too, so that people can rely on the makefile for all targets without exception.

vkjr commented:
@PombeirP, just checked: `make run-desktop` works fine on `develop` but doesn't work on `feature/nix`, yes.

PombeirP commented:
Thanks for checking, I'll take a look at what might have broken this.

PombeirP commented:
@vkjr interesting, I didn't change anything, but this time it worked (although I get a message that it could not open db):

![image](https://user-images.githubusercontent.com/138074/54464296-5e944880-4776-11e9-862f-e448f0d13d1c.png)

Commit:
Discard temporary changes to package.json in build-desktop.sh after compiling

rasom reviewed (approved):
tested local dev mode on iOS, Android and MacOS versions

asemiankevich commented:
@PombeirP could you please update us with the state of this PR ? @churik has tested it , also building this locally was done by developers. What else is left here ? cc @annadanchenko

PombeirP commented:
> @PombeirP could you please update us with the state of this PR ? @churik has tested it , also building this locally was done by developers. What else is left here ? cc @annadanchenko

We're almost ready to merge. The current state is:
    - most issues revolve around running react-native run-desktop. On some platforms it is not able to fetch the sources, and on Ubuntu 18.04 it is not able to decrypt the data while inside the nix-shell (this is also an issue seen when building the AppImage, but in that case we can exit the nix-shell and run the AppImage just fine). Not sure if we can merge before fixing this. cc @vkjr;
    - missing icons in MacOS: this is a known issue due to a bug in the `macdeployqt` which I’ll need to work around in our deployment script. Should be at most a day’s work, and we’ll do that before merging to develop so as to avoid a perceived broken nightly.

Right now I'll temporarily switch to the Chaos Unicorn Day effort and then later in the week get back to this.

churik commented:
@PombeirP I retested `.App` with https://github.com/status-im/status-react/commit/b8cbee917761065157e4cb3c4792f4e06a2d7627 on Linux 18.04.2 LTS (created new account, joining several chats, navigating through ). Looks good to me!

PombeirP commented:
@churik thanks! I'm working on getting icons working on macOS and should have it read after lunch (already have chat icons visible, just missing user account icon).

PombeirP commented:
@churik the latest macOS build should have the icons issue fixed

Commit:
Add ruby and bundler Nix packages for CI to fix android-e2e CI job

Commit:
Fix android debug build suffix

Signed-off-by: Pedro Pombeiro <pombeirp@users.noreply.github.com>

Serhy commented:
MacOS build validated on MacOS Mojave.

All icons are present just like with nightly build. 
By the way, MacOS .dmg is 25MB less here than Develop. Windows build 45MB less.
